File: services/parking-service/src/main/java/com/parkio/parking/presentation/dto/ApiError.java
package com.parkio.parking.presentation.dto;

import com.fasterxml.jackson.annotation.JsonInclude;
import java.time.Instant;
import java.util.List;
import org.slf4j.MDC;

/**
 * Consistent API error body (ai-context/04). {@code fieldErrors} is present only
 * for validation failures.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record ApiError(
        String code,
        String message,
        String traceId,
        List<FieldError> fieldErrors,
        Instant timestamp) {

    public record FieldError(String field, String message) {
    }

    public static ApiError of(String code, String message, Instant timestamp) {
        return new ApiError(code, message, currentTraceId(), null, timestamp);
    }

    public static ApiError of(String code, String message, List<FieldError> fieldErrors, Instant timestamp) {
        return new ApiError(code, message, currentTraceId(), fieldErrors, timestamp);
    }

    private static String currentTraceId() {
        String traceId = MDC.get("traceId");
        return traceId == null || traceId.isBlank() ? "unknown" : traceId;
    }
}

File: frontend/packages/validation/src/api-error.ts
import { z } from 'zod';
import type { ApiError, FieldError } from '@parkio/types';

export const fieldErrorSchema = z
  .object({
    field: z.string(),
    message: z.string(),
  })
  .strip() satisfies z.ZodType<FieldError>;

export const apiErrorSchema = z
  .object({
    code: z.string(),
    message: z.string(),
    traceId: z.string(),
    timestamp: z.string().datetime({ offset: true }),
    fieldErrors: z.array(fieldErrorSchema).optional(),
  })
  .strip() satisfies z.ZodType<ApiError>;

export type ApiErrorInput = z.infer<typeof apiErrorSchema>;
